java-checks/src/main/java/org/sonar/java/checks/NullShouldNotBeUsedWithOptionalCheck.java

```diff
@@ -0,0 +1,132 @@
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree method) {
+    if (!method.is(Tree.Kind.CONSTRUCTOR) && returnsOptional(method)) {
+
+      // check that the method is not annotated with @Nullable
+      ModifiersTree modifiers = method.modifiers();
+      for (AnnotationTree annotation : modifiers.annotations()) {
+        Type type = annotation.annotationType().symbolType();
+        if (type.is(NULLABLE)) {
+          context.reportIssue(this, annotation, "Methods with an \"Optional\" return type should not be \"@Nullable\".");
+        }
+      }
+
+      // check that the method does not return "null"
+      method.accept(new ReturnNullVisitor());
+    }
+
+    super.visitMethod(method);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree binaryExpression) {
+    String message = "Remove this null-check of an \"Optional\".";
+    if (binaryExpression.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+      ExpressionTree left = binaryExpression.leftOperand();
+      ExpressionTree right = binaryExpression.rightOperand();
+      if ((isOptional(left) && isNull(right)) || (isNull(left) && isOptional(right))) {
+        context.reportIssue(this, binaryExpression, message);
+      }
+    }
+
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  private class ReturnNullVisitor extends BaseTreeVisitor {
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree returnStatement) {
+      ExpressionTree expression = returnStatement.expression();
+      if (!checkNull(expression) && expression.is(Kind.CONDITIONAL_EXPRESSION)) {
+        ConditionalExpressionTree conditionalExpression = (ConditionalExpressionTree)expression; 
+        checkNull(conditionalExpression.trueExpression());
+        checkNull(conditionalExpression.falseExpression());
+      }
+    }
+    
+    private boolean checkNull(ExpressionTree expression) {
+      if (expression.is(Kind.NULL_LITERAL)) {
+        context.reportIssue(NullShouldNotBeUsedWithOptionalCheck.this, expression.firstToken(), "Methods with an \"Optional\" return type should never return null.");
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
+      // don't visit lambdas, as they are allowed to return null
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // don't visit subclasstree, as methods in there will be visited by outer class
+    }
+
+  }
+
+  private static boolean returnsOptional(MethodTree method) {
+    return method.returnType().symbolType().is(OPTIONAL);
+  }
+
+  private static boolean isOptional(ExpressionTree expression) {
+    return expression.symbolType().is(OPTIONAL) && !expression.is(Tree.Kind.NULL_LITERAL);
```

Comment:
`expression.is(Tree.Kind.NULL_LITERAL)` should be replaced by `isNull(expression)` in order to reuse the method.